lib/jxl/enc_external_image.cc

```diff
@@ -124,10 +125,19 @@ Status ConvertFromExternal(Span<const uint8_t> bytes, size_t xsize,
   // bits_per_sample > 1.
   const size_t bytes_per_pixel = DivCeil(bits_per_sample, jxl::kBitsPerByte);
 
-  const size_t row_size = xsize * bytes_per_pixel;
-  if (ysize && bytes.size() / ysize < row_size) {
+  const size_t last_row_size = xsize * bytes_per_pixel;
+  const size_t row_size =
+      (align > 1 ? jxl::DivCeil(last_row_size, align) * align : last_row_size);
+  const size_t bytes_to_read = row_size * (ysize - 1) + last_row_size;
+  if (xsize == 0 || ysize == 0) return JXL_FAILURE("Empty image");
+  if (bytes.size() < bytes_to_read) {
     return JXL_FAILURE("Buffer size is too small");
   }
+  // Too large buffer is likely an application bug, so also fail for that.
```

Comment:
I'm not sure I like this idea - I could imagine someone using a larger buffer which is virtually cropped and things getting more complicated. At the very least, I think this should happen in a separate PR.

Reply:
If you want to use a larger buffer and virtually crop it, you can always pass a smaller size than its actual size. This only works anyway if you want to crop away rows at the bottom — any other kind of crop requires you to do the math and compute the new pointer (to the crop origin), xsize /ysize and alignment (i.e. the original row_size).

Making it slightly easier to crop rows at the bottom (a rather specific thing) at the expense of losing an opportunity to catch API usage bugs doesn't seem to be worth it to me, but I agree that this is arguable so I'll make a separate PR for that.

Reply:
split this part out